Bug: [Pokemon Emerald/Core] - HM/Badge shuffle fails with unconnected entrances (KH1) if it has to swap

### What happened?

Pokemon Emerald with Randomize Badges/HMs set to Shuffle often fails to generate with KH1 present in the multiworld (or another world with unconnected entrances).

Pokemon Emerald performs its shuffled badge/hm fill in `connect_entrances`. It does not create the `CollectionState`s it uses with `allow_partial_entrances=True`, but it does use `fill_restrictive(single_player_placement=True)`.

Initially, it looks like Pokemon Emerald's HM/Badge shuffle fills should not be able to run into unconnected entrances (it does not have any unconnected entrances of its own), however, swap currently counts the reachable locations across the entire multiworld even with `single_player_placement=True`, which might be a bug in swap because most of `fill_restrictive` only checks the single-player world when `single_player_placement=True` is specified: https://github.com/ArchipelagoMW/Archipelago/blob/0.6.1/Fill.py#L140-L147

So, if Pokemon Emerald's `connect_entrances` is run first, a swap occurs in the fill for Pokemon Emerald's HM/Badge shuffle, and KH1 is present in the multiworld, generation will fail because swap will count reachable locations across the entire multiworld, which includes KH1 worlds with unconnected entrances.

I'm not sure if this is a Pokemon Emerald @Zunawe, Core, or both bug.

The `CollectionState`s created by Pokemon Emerald during `connect_entrances`:
https://github.com/ArchipelagoMW/Archipelago/blob/0.6.1/worlds/pokemon_emerald/__init__.py#L516
https://github.com/ArchipelagoMW/Archipelago/blob/0.6.1/worlds/pokemon_emerald/__init__.py#L555
The `fill_restrictive(single_player_placement=True)` calls Pokemon Emerald uses for the Badges/Hms shuffle fills:
https://github.com/ArchipelagoMW/Archipelago/blob/0.6.1/worlds/pokemon_emerald/__init__.py#L579

To reproduce, generate a few times with these yamls (default KH1 exported by archipelago.gg and default Pokemon Emerald exported by archipelago.gg, except `Randomize Badges` and `Randomize HMs` both set to `Shuffle`). Generation appears to fail much more often than it succeeds, so it should not take many tries to get a failing seed.
Note that the file names of the yamls are important so that Pokemon Emerald's `connect_entrances` is run before KH1's `connect_entrances`.
[Players.zip](https://github.com/user-attachments/files/19625001/Players.zip)

Example traceback from local generation on `main` (the same error can be produced by generating with these yamls on archipelago.gg):
```
Traceback (most recent call last):
  File "G:\git_Repos_other\Archipelago\Generate.py", line 573, in <module>
    multiworld = ERmain(erargs, seed)
                 ^^^^^^^^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\Main.py", line 154, in main
    AutoWorld.call_all(multiworld, "connect_entrances")
  File "G:\git_Repos_other\Archipelago\worlds\AutoWorld.py", line 194, in call_all
    call_single(multiworld, method_name, player, *args)
  File "G:\git_Repos_other\Archipelago\worlds\AutoWorld.py", line 184, in call_single
    raise e
  File "G:\git_Repos_other\Archipelago\worlds\AutoWorld.py", line 177, in call_single
    ret = _timed_call(method, *args, multiworld=multiworld, player=player)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\worlds\AutoWorld.py", line 163, in _timed_call
    ret = method(*args)
          ^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\worlds\pokemon_emerald\__init__.py", line 483, in connect_entrances
    self.shuffle_badges_hms()
  File "G:\git_Repos_other\Archipelago\worlds\pokemon_emerald\__init__.py", line 528, in shuffle_badges_hms
    self.fill_subset_with_retries(badge_items, badge_locations, state)
  File "G:\git_Repos_other\Archipelago\worlds\pokemon_emerald\__init__.py", line 580, in fill_subset_with_retries
    fill_restrictive(self.multiworld, state, locations_copy, items_copy, single_player_placement=True,
  File "G:\git_Repos_other\Archipelago\Fill.py", line 143, in fill_restrictive
    multiworld.get_reachable_locations(prev_state))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\BaseClasses.py", line 514, in get_reachable_locations
    return [location for location in self.get_locations(player) if location.can_reach(state)]
                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\BaseClasses.py", line 1303, in can_reach
    return self.parent_region.can_reach(state) and self.access_rule(state)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\git_Repos_other\Archipelago\BaseClasses.py", line 1179, in can_reach
    state.update_reachable_regions(self.player)
  File "G:\git_Repos_other\Archipelago\BaseClasses.py", line 755, in update_reachable_regions
    self._update_reachable_regions_explicit_indirect_conditions(player, queue)
  File "G:\git_Repos_other\Archipelago\BaseClasses.py", line 771, in _update_reachable_regions_explicit_indirect_conditions
    assert new_region, f"tried to search through an Entrance \"{connection}\" with no connected Region"
           ^^^^^^^^^^
AssertionError: tried to search through an Entrance "Awakening (KH1Default)" with no connected Region
Exception in <bound method PokemonEmeraldWorld.connect_entrances of <worlds.pokemon_emerald.PokemonEmeraldWorld object at 0x000001BD14FBFE90>> for player 1, named BadgeHmShuffle.
```

### What were the expected results?

Generation should succeed.

### Software

Website

--- a/Fill.py
+++ b/Fill.py
@@ -138,32 +138,21 @@ def fill_restrictive(multiworld: MultiWorld, base_state: CollectionState, locati
                         # to clean that up later, so there is a chance generation fails.
                         if (not single_player_placement or location.player == item_to_place.player) \
                                 and location.can_fill(swap_state, item_to_place, perform_access_check):
+                            # Add this item to the existing placement, and
+                            # add the old item to the back of the queue
+                            spot_to_fill = placements.pop(i)
 
-                            # Verify placing this item won't reduce available locations, which would be a useless swap.
-                            prev_state = swap_state.copy()
-                            prev_loc_count = len(
-                                multiworld.get_reachable_locations(prev_state))
+                            swap_count += 1
+                            swapped_items[placed_item.player, placed_item.name, unsafe] = swap_count
 
-                            swap_state.collect(item_to_place, True)
-                            new_loc_count = len(
-                                multiworld.get_reachable_locations(swap_state))
+                            reachable_items[placed_item.player].appendleft(
+                                placed_item)
+                            item_pool.append(placed_item)
 
-                            if new_loc_count >= prev_loc_count:
-                                # Add this item to the existing placement, and
-                                # add the old item to the back of the queue
-                                spot_to_fill = placements.pop(i)
+                            # cleanup at the end to hopefully get better errors
+                            cleanup_required = True
 
-                                swap_count += 1
-                                swapped_items[placed_item.player, placed_item.name, unsafe] = swap_count
-
-                                reachable_items[placed_item.player].appendleft(
-                                    placed_item)
-                                item_pool.append(placed_item)
-
-                                # cleanup at the end to hopefully get better errors
-                                cleanup_required = True
-
-                                break
+                            break
 
                         # Item can't be placed here, restore original item
                         location.item = placed_item

